tiled/server/pydantic_array.py

```diff
@@ -39,10 +41,18 @@ class BuiltinDtype(BaseModel):
 
     @classmethod
     def from_numpy_dtype(cls, dtype) -> "BuiltinDtype":
+        # Extract datetime units from the dtype string representation,
+        # e.g. `'<M8[ns]'` has `units = 'ns'`.
+        units = ""
+        if dtype.kind in ("m", "M"):
+            if res := re.search(r"\[(.+)\]$", dtype.str):
```

Comment:
Surely there is a better way than running a regex on the string version of the dtype ?!

Reply:
If we can not dig this out, I think it would be better to go with `units = {dtype_obj: 'string', ..}[dtype]` where we directly map the numpy dtype to the string we want for it.  A bit of boilerplate for us, but it is clear what is going on and protects us from most of the things I can think of that numpy could change that would mess with us (and it does not user their str repr as a critcal API).

Reply:
This would mean we could also pick some units _not_ to support https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units

Reply:
Turns out, there is a helper [utility function](https://numpy.org/doc/stable/reference/generated/numpy.datetime_data.html) in numpy to extract the units from `datetime` dtypes, and the units can be even more complicated. Thank you, @danielballan for pointing this out! This is probably the cleanest solution, and I hope numpy wouldn't change this api. I don't think we need to keep the units (strings) and counts (int) separately, though -- a compound attribute is enough -- but keen to hear your thoughts on this.

Reply:
That is something I would expect them to keep stable.